crates/cuda-common/src/memory_manager/tests.rs

```diff
@@ -0,0 +1,505 @@
+// Cross-thread: Thread A frees, Thread B should be able to reuse after sync
+// ============================================================================
+#[test]
+fn test_cross_thread_handoff() {
+    let runtime = tokio::runtime::Builder::new_multi_thread()
+        .worker_threads(2)
+        .max_blocking_threads(2)
+        .enable_all()
+        .build()
+        .unwrap();
+
+    runtime.block_on(async {
+        let barrier = Arc::new(Barrier::new(2));
+        // Store address as usize (which is Send + Sync)
+        let addr_holder = Arc::new(std::sync::Mutex::new(0usize));
+
+        let barrier1 = barrier.clone();
+        let addr_holder1 = addr_holder.clone();
+
+        // Thread A: allocate and free
+        let handle_a = tokio::task::spawn(async move {
+            tokio::task::spawn_blocking(move || {
+                let len = (200 << 20) / size_of::<u8>(); // 200 MB
+                let buf = DeviceBuffer::<u8>::with_capacity(len);
+                *addr_holder1.lock().unwrap() = buf.as_raw_ptr() as usize;
+                // buf drops here, freeing memory
+            })
+            .await
+            .expect("task A panicked");
+
+            barrier1.wait().await;
+        });
+
+        let barrier2 = barrier.clone();
+        let addr_holder2 = addr_holder.clone();
+
+        // Thread B: wait for A, then allocate same size
+        let handle_b = tokio::task::spawn(async move {
+            barrier2.wait().await;
+
+            tokio::task::spawn_blocking(move || {
+                let len = (200 << 20) / size_of::<u8>();
+                let buf = DeviceBuffer::<u8>::with_capacity(len);
+
+                // Check if we got the same address (depends on event sync timing)
+                let original_addr = *addr_holder2.lock().unwrap();
+                let new_addr = buf.as_raw_ptr() as usize;
+                if new_addr == original_addr {
+                    println!("Cross-thread: reused same address (event synced)");
+                } else {
+                    println!("Cross-thread: different address (event pending)");
+                }
+                // buf drops here
+            })
+            .await
+            .expect("task B panicked");
+        });
+
+        handle_a.await.expect("thread A failed");
+        handle_b.await.expect("thread B failed");
+    });
+}
+
+// ============================================================================
+// Stress: many threads doing random alloc/free patterns
+// ============================================================================
+#[test]
+fn test_stress_multithread() {
+    let runtime = tokio::runtime::Builder::new_multi_thread()
+        .worker_threads(8)
+        .max_blocking_threads(8)
+        .enable_all()
+        .build()
+        .unwrap();
+
+    runtime.block_on(async {
+        let mut handles = Vec::new();
+
+        for thread_idx in 0..8 {
+            let handle = tokio::task::spawn_blocking(move || {
```

Comment:
what is this testing? if it's not close to any hard memory limits, the behavior could be incorrect but the allocations still work. I expect more asserts of exact behavior? Right now even if something didn't go correctly, since there's no kernels accessing the memory, I doubt you'd catch much with this test